src/pymoca/tree.py

```diff
@@ -747,6 +775,494 @@ def _check_import_rules(
         raise NameLookupError(f"Import {full_name} is recursive")
 
 
+class InstanceTree(ast.Tree):
+    """
+    The root class of an instance tree
+    """
+
+    BUILTIN_TYPES = {
+        "Real": ast.Symbol(
+            name="Real",
+            type=ast.ComponentRef(name="Real"),
+            start=ast.Primary(value=0.0),
+            min=ast.Primary(value=-math.inf),
+            max=ast.Primary(value=math.inf),
+            nominal=ast.Primary(value=None),
+            fixed=ast.Primary(value=False),  # True for parameters and constants
+            unit=ast.Primary(value=""),
+            quantity=ast.Primary(value=""),
+            displayUnit=ast.Primary(value=""),
+            # TODO: unbounded from spec is missing in Symbol
+            # TODO: stateSelect from spec is missing in Symbol
+        ),
+        "Integer": ast.Symbol(
+            name="Integer",
+            type=ast.ComponentRef(name="Integer"),
+            start=ast.Primary(value=0.0),
+            min=ast.Primary(value=-sys.maxsize),
+            max=ast.Primary(value=sys.maxsize),
+            fixed=ast.Primary(value=False),  # True for parameters and constants
+            quantity=ast.Primary(value=""),
+        ),
+        "Boolean": ast.Symbol(
+            name="Boolean",
+            type=ast.ComponentRef(name="Boolean"),
+            start=ast.Primary(value=0.0),
+            fixed=ast.Primary(value=False),  # True for parameters and constants
+            quantity=ast.Primary(value=""),
+        ),
+        "String": ast.Symbol(
+            name="String",
+            type=ast.ComponentRef(name="String"),
+            start=ast.Primary(value=0.0),
+            fixed=ast.Primary(value=False),  # True for parameters and constants
+            quantity=ast.Primary(value=""),
+        ),
+    }
+
+    def __init__(self, ast_ref: ast.Tree, **kwargs):
+        # The Class AST
+        self.ast_ref = ast_ref
+
+        super().__init__(**kwargs)
+        self.create_builtins()
+
+    def create_builtins(self):
+        """Add builtins to root of tree"""
+        for name, symbol in self.BUILTIN_TYPES.items():
+            type_class = ast.Class(name=name, type=name, parent=self)
+            symbol.parent = type_class
+            type_class.symbols[name] = symbol
+            self.classes[name] = type_class
+        # TODO: Add built-in functions (and operators?)
+
+    def instantiate(self, class_name: str) -> ast.InstanceClass:
+        """Create an instance tree used in flattening"""
+        class_ = find_name(class_name, self.ast_ref)
```

Comment:
No, that looks like a bug! Need a test case!

Reply:
I don't see anything in the spec related to this, so I think this is more of an tool implementation decision than a Modelica language restriction.

Here's an example:

```Modelica
package A
    encapsulated model B
        constant Boolean x = true;
    end B;
    model C
        import A.B;
        B b;
    end C;
end A;
```

The current implementation will instantiate `A.C.B`, with the resulting `InstanceClass` having `A` as a parent, which is a little surprising. If we instead instantiate `A.C`, then `A.C` symbol `b` `type.parent` is also `A`, so at least it is consistent. 

I think if we were to do a `search_imports=False` (using `_find_name`), then it would be passed along the call stack and mess up the logic in recursive calls, so we would instead need to add complication to `instantiate` to check it up front.

I don't know if it really should be prevented because from a language standpoint once the import is stated, it becomes part of the namespace of the containing class, so it *is* valid to refer to it. For an edge case like this, I don't think it is worth the complication to change the current behavior.

Reply:
@jackvreeken I'll let you comment or resolve this one.

Reply:
Sounds fair enough to me, I just figured it'd be a little weird to e.g. flatten `Modelica.Units` using a totally different model that just imports it. But indeed, just because it's weird might not mean it's wrong :)